src/daft-logical-plan/src/optimization/rules/granular_projections.rs

```diff
@@ -0,0 +1,389 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use common_treenode::{Transformed, TreeNode};
+use daft_dsl::{functions::ScalarFunction, Column, Expr, ResolvedColumn};
+use itertools::Itertools;
+
+use super::OptimizerRule;
+use crate::{ops::Project, LogicalPlan};
+
+/// This rule will split projections into multiple projections such that expressions that
+/// need their own granular morsel sizing will be isolated. Right now, those would be
+/// URL downloads, but this may be extended in the future to other functions and Python UDFs.
+#[derive(Debug)]
+pub struct GranularProjections {}
```

Comment:
Our naming convention for optimizer rules seems to be to have some sort of singular verb. Maybe `SplitGranularProjection`?